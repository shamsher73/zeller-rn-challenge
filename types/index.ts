export type User = {
    id: string,
    name: string,
    role: string,
}

export enum UserType {
    ADMIN = 'ADMIN',
    MANAGER = 'MANAGER',
}

export type QueryResponse = {
    data: {
        listZellerCustomers: {
            items: User[];
        };
    }
};