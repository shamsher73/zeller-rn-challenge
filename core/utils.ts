import { Customer } from "@/types";

export function capitalizeWords(str: string) {
    return str
        .toLowerCase()
        .split(' ')
        .map(word => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
}

export function filterItemsWithRole(items: Customer[], selectedRole: string): Customer[] {
    return items.filter(item => item.role.toLowerCase() === selectedRole.toLowerCase());
}

export function filterItemsWithNameRole(items: Customer[], selectedRole: string, name: string): Customer[] {
    return items.filter(item => item.name.toLowerCase().includes(name.toLowerCase()) &&
        item.role.toLowerCase() === selectedRole.toLowerCase());
}