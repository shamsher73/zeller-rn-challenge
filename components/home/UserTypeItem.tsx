import { StyleSheet, TouchableOpacity } from 'react-native';
import { ThemedText } from '../ThemedText';
import MaterialIcons from '@expo/vector-icons/MaterialIcons';
import { Colors } from '@/constants/Colors';
import { UserType } from '@/types';
import { capitalizeWords } from '@/core/utils';

type UserTypeItemProps = {
    title: UserType,
    selectedValue: UserType,
    setSelectedValue: (item: UserType) => void
}

export default function UserTypeItem(props: UserTypeItemProps) {
    const { title, selectedValue, setSelectedValue } = props
    const isSelected = selectedValue === title ? true : false;
    return (
        <TouchableOpacity
            style={[styles.container, { backgroundColor: isSelected ? Colors.secondary : Colors.background }]}
            onPress={() => setSelectedValue(title)}
            testID='user-type-item'
        >
            <MaterialIcons
                name={isSelected ? "radio-button-checked" : "radio-button-unchecked"}
                size={24}
                color={Colors.primary}
                style={styles.icon}
            />
            <ThemedText >{capitalizeWords(title)}</ThemedText>
        </TouchableOpacity>
    );
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingVertical: 12,
        borderRadius: 8,
    },
    icon: {
        marginHorizontal: 10
    }
});
