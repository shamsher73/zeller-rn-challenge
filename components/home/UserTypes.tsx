import { StyleSheet, View } from 'react-native';
import UserTypeItem from './UserTypeItem';
import { ThemedText } from '../ThemedText';
import { UserType } from '@/types';
import Divider from '../Divider';

type UserTypesProps = {
    list: UserType[],
    selectedValue: UserType,
    setSelectedValue: (title: UserType) => void
}

export default function UserTypes(props: UserTypesProps) {
    const { list, selectedValue, setSelectedValue } = props
    const UserList = () => {
        return list.map((item, index) => {
            return <UserTypeItem key={index} title={item} selectedValue={selectedValue} setSelectedValue={setSelectedValue} />
        })
    }

    return (
        <View>
            <ThemedText type='title'>User Types</ThemedText>
            <View style={styles.listContainer} >
                {UserList()}
            </View>
            <Divider />
        </View>
    );
}

const styles = StyleSheet.create({
    listContainer: {
        marginVertical: 20
    }
});
