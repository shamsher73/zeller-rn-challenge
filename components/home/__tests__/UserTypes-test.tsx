import { UserType } from '@/types';
import { render } from '@testing-library/react-native';
import UserTypeItem from '../UserTypeItem';
import UserTypes from '../UserTypes';
import { capitalizeWords } from '@/utils';

describe('<UserTypes />', () => {
    it('should render UserTypes component correctly with given props', () => {
        const list = [UserType.ADMIN, UserType.MANAGER]

        const selectedValue = UserType.ADMIN;
        const setSelectedValue = jest.fn();

        const { getByText } = render(
            <UserTypes list={list} selectedValue={selectedValue} setSelectedValue={setSelectedValue} />
        );

        expect(getByText('User Types')).toBeTruthy();
        list.forEach(item => {
            expect(getByText(capitalizeWords(item))).toBeTruthy();
        });
    });

    it('should handle an empty list without errors', () => {
        const list: UserType[] = []
        const selectedValue = UserType.ADMIN;
        const setSelectedValue = jest.fn();

        const { getByText, queryByText } = render(
            <UserTypes list={list} selectedValue={selectedValue} setSelectedValue={setSelectedValue} />
        );

        expect(getByText('User Types')).toBeTruthy();
        expect(queryByText('Admin')).toBeNull();
        expect(queryByText('Manager')).toBeNull();
    });

    test('snapshot', () => {
        const list = [UserType.ADMIN, UserType.MANAGER]
        const selectedValue = UserType.ADMIN;
        const setSelectedValue = jest.fn();
        const tree = render(<UserTypes list={list} selectedValue={selectedValue} setSelectedValue={setSelectedValue} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
