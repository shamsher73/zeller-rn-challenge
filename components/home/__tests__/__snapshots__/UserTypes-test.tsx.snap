// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserTypes /> snapshot 1`] = `
<View>
  <Text
    style={
      [
        undefined,
        {
          "fontSize": 22,
          "fontWeight": "500",
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    User Types
  </Text>
  <View
    style={
      {
        "marginVertical": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#e8f2fb",
          "borderRadius": 8,
          "flexDirection": "row",
          "opacity": 1,
          "paddingVertical": 12,
        }
      }
      testID="user-type-item"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#0370ce",
              "fontSize": 24,
            },
            {
              "marginHorizontal": 10,
            },
            {
              "fontFamily": "material",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
            },
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        Admin
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "flexDirection": "row",
          "opacity": 1,
          "paddingVertical": 12,
        }
      }
      testID="user-type-item"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#0370ce",
              "fontSize": 24,
            },
            {
              "marginHorizontal": 10,
            },
            {
              "fontFamily": "material",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
            },
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        Manager
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "borderBottomColor": "grey",
        "borderBottomWidth": 0.5,
        "marginVertical": 16,
      }
    }
  />
</View>
`;
